jupyter lab build  fails in webpack install

I'm trying to install jupyter lab extensions in a clean linux box.  I'm getting install errors which appear to occur because jupyter lab build is calling webpack and not webpack --dev so that the devdependencies are not getting installed

warning ipysheet > css-loader > cssnano > postcss-merge-rules > caniuse-api > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
[2/4] Fetching packages...
[###########################----------------------------------------] 419/1051[LabBuildApp] > node /usr/lib/python3.6/site-packages/jupyterlab/staging/yarn.js run build
yarn run v1.9.4
$ webpack
✖ ｢config｣: An error occurred while trying to find a config file
              Did you forget to specify a --require?
/usr/lib/node_modules/webpack-command/node_modules/@webpack-contrib/config-loader/lib/load.js:85
    throw new LoadConfigError(e, configPath);
    ^

LoadConfigError: Cannot find module 'fs-extra'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (/usr/lib/node_modules/webpack-command/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
    at Object.<anonymous> (/usr/share/jupyter/lab/staging/webpack.config.js:7:10)
    at Module._compile (/usr/lib/node_modules/webpack-command/node_modules/v8-compile-cache/v8-compile-cache.js:178:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

I've been having the same issue, but with an additional line about the registry:
```
yarn install v1.9.4
info No lockfile found.
[1/4] Resolving packages...
error Received malformed response from registry for undefined. The registry may be down.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
[LabBuildApp] > node C:\cleanconda\envs\ipycolordev\lib\site-packages\jupyterlab\staging\yarn.js run build
yarn run v1.9.4
$ webpack
module.js:538
    throw err;
    ^

Error: Cannot find module 'fs-extra'
    [...]
```

--- a/jupyterlab/build_handler.py
+++ b/jupyterlab/build_handler.py
@@ -89,7 +89,7 @@ class Builder(object):
 
     @run_on_executor
     def _run_build(self, app_dir, logger, kill_event):
-        kwargs = dict(app_dir=app_dir, logger=logger, kill_event=kill_event)
+        kwargs = dict(app_dir=app_dir, logger=logger, kill_event=kill_event, command='build')
         try:
             return build(**kwargs)
         except Exception as e:

--- a/jupyterlab/commands.py
+++ b/jupyterlab/commands.py
@@ -1480,8 +1480,9 @@ class _AppHandler(object):
 def _node_check(logger):
     """Check for the existence of nodejs with the correct version.
     """
+    node = which('node')
     try:
-        output = subprocess.check_output(['node', 'node-version-check.js'], cwd=HERE)
+        output = subprocess.check_output([node, 'node-version-check.js'], cwd=HERE)
         logger.info(output.decode('utf-8'))
     except Exception:
         data = _get_core_data()

--- a/jupyterlab/commands.py
+++ b/jupyterlab/commands.py
@@ -15,6 +15,7 @@ import os.path as osp
 import re
 import shutil
 import site
+import subprocess
 import sys
 import tarfile
 from tempfile import TemporaryDirectory
@@ -150,7 +151,7 @@ def watch_packages(logger=None):
         yarn_proc = Process(['node', YARN_PATH], cwd=parent, logger=logger)
         yarn_proc.wait()
 
-    logger = logger or logging.getLogger('jupyterlab')
+    logger = _ensure_logger(logger)
     ts_dir = osp.realpath(osp.join(HERE, '..', 'packages', 'metapackage'))
 
     # Run typescript watch and wait for the string indicating it is done.
@@ -178,7 +179,7 @@ def watch_dev(logger=None):
     -------
     A list of `WatchHelper` objects.
     """
-    logger = logger or logging.getLogger('jupyterlab')
+    logger = _ensure_logger(logger)
 
     package_procs = watch_packages(logger)
 
@@ -204,7 +205,8 @@ def watch(app_dir=None, logger=None):
     -------
     A list of processes to run asynchronously.
     """
-    _node_check()
+    logger = _ensure_logger(logger)
+    _node_check(logger)
     handler = _AppHandler(app_dir, logger)
     return handler.watch()
 
@@ -216,7 +218,8 @@ def install_extension(extension, app_dir=None, logger=None):
 
     Returns `True` if a rebuild is recommended, `False` otherwise.
     """
-    _node_check()
+    logger = _ensure_logger(logger)
+    _node_check(logger)
     handler = _AppHandler(app_dir, logger)
     return handler.install_extension(extension)
 
@@ -226,7 +229,8 @@ def uninstall_extension(name, app_dir=None, logger=None):
 
     Returns `True` if a rebuild is recommended, `False` otherwise.
     """
-    _node_check()
+    logger = _ensure_logger(logger)
+    _node_check(logger)
     handler = _AppHandler(app_dir, logger)
     return handler.uninstall_extension(name)
 
@@ -239,7 +243,8 @@ def update_extension(name=None, all_=False, app_dir=None, logger=None):
 
     Returns `True` if a rebuild is recommended, `False` otherwise.
     """
-    _node_check()
+    logger = _ensure_logger(logger)
+    _node_check(logger)
     handler = _AppHandler(app_dir, logger)
     if all_ is True:
         return handler.update_all_extensions()
@@ -264,7 +269,8 @@ def build(app_dir=None, name=None, version=None, public_url=None,
           clean_staging=False):
     """Build the JupyterLab application.
     """
-    _node_check()
+    logger = _ensure_logger(logger)
+    _node_check(logger)
     handler = _AppHandler(app_dir, logger, kill_event=kill_event)
     return handler.build(name=name, version=version, public_url=public_url,
                          command=command, clean_staging=clean_staging)
@@ -307,7 +313,8 @@ def build_check(app_dir=None, logger=None):
 
     Returns a list of messages.
     """
-    _node_check()
+    logger = _ensure_logger(logger)
+    _node_check(logger)
     handler = _AppHandler(app_dir, logger)
     return handler.build_check()
 
@@ -377,7 +384,7 @@ class _AppHandler(object):
         """
         self.app_dir = app_dir or get_app_dir()
         self.sys_dir = get_app_dir()
-        self.logger = logger or logging.getLogger('jupyterlab')
+        self.logger = _ensure_logger(logger)
         self.info = self._get_app_info()
         self.kill_event = kill_event or Event()
         # TODO: Make this configurable
@@ -1470,19 +1477,22 @@ class _AppHandler(object):
         return proc.wait()
 
 
-def _node_check():
+def _node_check(logger):
     """Check for the existence of nodejs with the correct version.
     """
     try:
-        proc = Process(['node', 'node-version-check.js'], cwd=HERE, quiet=True)
-        ret = proc.wait()
+        output = subprocess.check_output(['node', 'node-version-check.js'], cwd=HERE)
+        logger.info(output.decode('utf-8'))
     except Exception:
-        ret = 1
-    if ret != 0:
         data = _get_core_data()
         ver = data['engines']['node']
         msg = 'Please install nodejs %s before continuing. nodejs may be installed using conda or directly from the nodejs website.' % ver
         raise ValueError(msg)
+
+
+def _ensure_logger(logger=None):
+    """Ensure that we have a logger"""
+    return logger or logging.getLogger('jupyterlab')
 
 
 def _normalize_path(extension):

--- a/jupyterlab/labapp.py
+++ b/jupyterlab/labapp.py
@@ -63,10 +63,16 @@ class LabBuildApp(JupyterApp):
     dev_build = Bool(True, config=True,
         help="Whether to build in dev mode (defaults to dev mode)")
 
+    pre_clean = Bool(True, config=True,
+        help="Whether to clean before building (defaults to True)")
+
     def start(self):
         command = 'build:prod' if not self.dev_build else 'build'
         app_dir = self.app_dir or get_app_dir()
         self.log.info('JupyterLab %s', version)
+        if self.pre_clean:
+            self.log.info('Cleaning %s' % app_dir)
+            clean(self.app_dir)
         self.log.info('Building in %s', app_dir)
         build(app_dir=app_dir, name=self.name, version=self.version,
               command=command, logger=self.log)

--- a/jupyterlab/node-version-check.js
+++ b/jupyterlab/node-version-check.js
@@ -11,6 +11,9 @@ var eparts = engine.split('.').map(parser);
 var version = process.version.replace('v', '');
 var vparts = version.split('.').map(parser);
 
+// eslint-disable-next-line
+console.log('Node', process.version);
+
 if (vparts[0] > eparts[0]) {
   process.exit(0);
 }

